### PR TITLE
Measurements command

### PR DESCRIPTION
**Pull request by joverlee521**
### Description of proposed changes
Adds the `measurements` subcommand with two sub-subcommands, `export` and `concat`.
See commits for details. 

### Related issue(s)
Based on proposal in #869 

### Testing
Added unit tests for the changes in the `validate` subcommand. 
Added functional tests for `measurements export` and `measurements concat`.

### CODE REVIEW

**Review by huddlej — CHANGES_REQUESTED** [review]
Really nice work, @joverlee521! This is probably the most thoroughly documented and tested a new Augur subcommand has ever been. The implementation is generally quite clear and easy to follow. The API docs and tests go a long way toward demonstrating how to use this new command. We might eventually want a how-to guide or tutorial for using this, but that's better for a later PR.

Although I've requested changes, my comments below are primarily stylistic. The exception are the comments about how functions are defined in the new `measurements.py` module (which are more about function than style). We can chat about these comments more synchronously with anyone else who has feedback on the command.

File: augur/measurements.py [hunk 1]
```diff
@@ -0,0 +1,364 @@
...
+
+    Returns
+    -------
+    bool
+        True if column exists in collection
+    """
+    column_in_df = column in collection.columns
+    if not column_in_df:
+        print(
+            f"ERROR: Provided {column_purpose} column «{column}» does not exist in collection TSV.",
```
[the remark this answers is not in the document]
Comment (tsibley) [reply augur/measurements.py:47]: A nice technique for this is to rely on Python's built-in `repr()` formatting, which you can succinctly use within f-strings by specifying a `!r` conversion like so:

```python
f"ERROR: Provided {column_purpose} column {column!r} …`
```

This improves on:

```python
f"ERROR: Provided {column_purpose} column '{column}' …`
```

since `repr()` unambiguously represents single quotes and other potentially confusing chars in the value whereas the latter doesn't.

See the [f-string spec](https://docs.python.org/3/reference/lexical_analysis.html#f-strings) for details on the conversion flags.

[the remark this answers is not in the document]
Comment (tsibley) [reply augur/measurements.py:47]: (The use of guillemets or "French quotes" («») for surrounding literals in error messages probably comes from my habit of using them, which arose from wanting unambiguous (well, _less_ ambiguous) quotes in programming langs that don't have something like `repr()` close at hand. It's a hack that relies on guillemets not appearing in the value and reintroducing ambiguity, which obviously is very context dependent! `repr()` is better here. This is all to say, sorry if I've lent you this habit, Jover!)

[the remark this answers is not in the document]
Comment (huddlej) [reply augur/measurements.py:47]: Good call about using `!r` as standard practice, @tsibley!

Also, I had to look up guillemets and its pronunciation, but I get where you're coming from now! :) I've learned a lot from this PR already...

File: augur/measurements.py [hunk 2]
```diff
@@ -0,0 +1,364 @@
...
+             "field in the config JSON will be dropped.")
+    config.add_argument("--key",
+        help="A short key name of the collection for internal use within auspice." +
+             "If not provided via config or command line option, the collection TSV filename will be used. ")
+    config.add_argument("--title",
+        help="The full title of the collection to display in the measurements panel title. " +
+             "If not provided via config or command line option, the panel's default title is 'Measurements'.")
+    config.add_argument("--x-axis-label",
+        help="The short label to display for the x-axis that describles the value of the measurements. " +
+             "If not provided via config or command line option, the panel's default x-axis label is 'measurements values'.")
```
[the remark this answers is not in the document]
Comment (joverlee521) [reply augur/measurements.py:323]: Ah, as you stated in another [comment](https://github.com/nextstrain/augur/pull/879#discussion_r853568013), I don't want to always overwrite the config file with defaults. String mismatches within the help message are not great either. I can make a global `defaults` dict to hold these values.

File: augur/measurements.py [hunk 3]
```diff
@@ -0,0 +1,364 @@
...
+    config.add_argument("--threshold", type=float,
```
[the remark this answers is not in the document]
Comment (joverlee521) [reply augur/measurements.py:324]: Yeah, I also see this changing in the near future. I see your point of the detailed appearance description being too much to maintain. I can reword it to something more generic.

File: augur/measurements.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (joverlee521) [reply augur/measurements.py:47]: Sorry, definitely a habit carried over from SFS days 😅 
I've seen the use of `!r` in other PRs, time to pick up a new habit.

File: augur/measurements.py [hunk 4]
```diff
@@ -0,0 +1,364 @@
...
+class HideAsFalseAction(argparse.Action):
+    """
+    Custom argparse Action that stores False for arguments passed as `--hide*`
+    and stores True for all other argument patterns.
+    """
+    def __call__(self, parser, namespace, values, option_string=None):
+        setattr(namespace, self.dest, option_string[2:6] != 'hide')
+
+
+def column_exists(collection, column, column_purpose):
```
[the remark this answers is not in the document]
Comment (joverlee521) [reply augur/measurements.py:25]: Ah, this is another habit from SFS days. I'm used to corralling generic QC checks like this into separate functions that can be reused. I'll remove the separate function.

File: augur/measurements.py [hunk 5]
```diff
@@ -0,0 +1,364 @@
...
+        collection_df = pd.read_csv(collection, sep="\t")
+    except FileNotFoundError:
+        print(
+            f"ERROR: collection TSV file ({collection}) does not exist",
+            file=sys.stderr,
+        )
+        return None
+
+    # Verify the strain and value columns are different
+    if strain_column == value_column:
```
[the remark this answers is not in the document]
Comment (joverlee521) [reply augur/measurements.py:85]: Definitely not expecting this to be a common issue, I just ran into this when I made a copy/paste error in testing

File: augur/validate.py [hunk 6]
```diff
@@ -119,13 +119,175 @@ def export_v1(meta_json, tree_json, **kwargs):
...
     return measurements
 
 
+def measurements_collection_config(collection_config_json, **kwargs):
+    schema = load_json_schema("schema-measurements-collection-config.json")
+    collection_config = load_json(collection_config_json)
+    validate_json(collection_config, schema, collection_config_json)
+    if not validate_collection_display_defaults(collection_config):
+        raise ValidateError("Validation of the collection config display defaults failed.")
+    return collection_config
```
[the remark this answers is not in the document]
Comment (joverlee521) [reply augur/validate.py:288]: It's not immediately obvious, but this function gets called as the sub-subcommand `augur validate measurements-collection-config`, through: 
 https://github.com/nextstrain/augur/blob/a923e39537e9b327e30d2c2199e5671a7d90bb2b/augur/validate.py#L147

File: augur/measurements.py [hunk 7]
```diff
@@ -0,0 +1,364 @@
...
+    ValidateError
+)
+
+class HideAsFalseAction(argparse.Action):
+    """
+    Custom argparse Action that stores False for arguments passed as `--hide*`
+    and stores True for all other argument patterns.
+    """
+    def __call__(self, parser, namespace, values, option_string=None):
+        setattr(namespace, self.dest, option_string[2:6] != 'hide')
```
[the remark this answers is not in the document]
Comment (joverlee521) [reply augur/measurements.py:22]: Sounds good, I wasn't sure if it would have any wider uses outside of this command.

File: augur/measurements.py [hunk 8]
```diff
@@ -0,0 +1,364 @@
...
+def register_arguments(parser):
+    subparsers = parser.add_subparsers(dest='subcommand')
+    subparsers.required = True
+
+    export = subparsers.add_parser("export", help="Export a measurements JSON for a single collection")
+
+    required = export.add_argument_group(
+        title="REQUIRED"
+    )
+    required.add_argument("--collection", required=True, metavar="TSV",
```
[the remark this answers is not in the document]
Comment (joverlee521) [reply augur/measurements.py:285]: Good point for the flu builds. I can add a new `--include-columns` option to get a subset of columns from the TSV. If not provided, the default behavior will still be pull in all columns. 

I will also update error messages for missing columns to specify if error arises from column not being in the TSV or from column not being listed for `--include-columns`.

File: augur/measurements.py [hunk 9]
```diff
@@ -0,0 +1,364 @@
...
+def concat_measurements(args):
+    output = {
+        'collections': []
+    }
+    if args.get("default_collection"):
+        output['default_collection'] = args['default_collection']
+
+    for json in args['jsons']:
+        measurements = validate_measurements_json(json)
+        output['collections'].extend(measurements['collections'])
```
[the remark this answers is not in the document]
Comment (joverlee521) [reply augur/measurements.py:269]: I would push this to later improvements since it technically doesn't break anything. I can see this as a part of the work to add measurements data as color-by's to the tree.

File: augur/validate.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (huddlej) [reply augur/validate.py:288]: Wow...thank you for pointing that out! I take back anything mean I've ever said about Perl syntax... :D

File: augur/measurements.py [hunk 10]
```diff
@@ -0,0 +1,364 @@
...
+             "If not provided, the first collection of the first JSON file will be displayed")
+    concat.add_argument("--minify-json", action="store_true",
+        help="Concat JSONs without indentation or line returns.")
+    concat.add_argument("--output-json", required=True, metavar="JSON", type=str,
+        help="Output JSON file")
+
+
+def run(args):
+    if args.subcommand == 'export':
+        return export_measurements(vars(args))
```
[the remark this answers is not in the document]
Comment (joverlee521) [reply augur/measurements.py:362]: I used `vars(args)` because it's used in the validate module. I see now that it's being used as `**vars(args)` [in the validate module](https://github.com/nextstrain/augur/blob/8014186d35c13d9f131cfa4828c6b7f81932909f/augur/validate.py#L147) to pass args as keyword arguments to the subcommand functions. 

I can change to passing the `args` object directly to fit in with other Augur subcommands.

File: augur/measurements.py [hunk 11]
```diff
@@ -0,0 +1,364 @@
...
+        validate_measurements_json(measurements_json=output_json)
+    except ValidateError:
+        print(
+            "ERROR: Validation of output JSON failed. See detailed errors above.",
+            file=sys.stderr,
+        )
+        sys.exit(1)
+
+
+def export_measurements(args):
```
[the remark this answers is not in the document]
Comment (joverlee521) [reply augur/measurements.py:201]: I've been mulling over this comment and keep waffling on it...

On one hand, I want to push back on moving all of the functions into the "main" function. I think the functions can exist even if we don't plan on reusing them. I view the functions as self-contained bits of logic that can be easier to understand than having to read through the entire "main" function. 

On the other hand, this would be such a straightforward "main" function that I can see the simplicity of moving everything. Then we can avoid unnecessary raising/excepting within functions. 

All to say, I'm not sure what's the best thing to do here 😅  
I can add a final "overhaul" commit that moves everything into "main" to see what it looks like after addressing all of the other comments.

File: augur/measurements.py [hunk 12]
```diff
@@ -0,0 +1,364 @@
...
+        sys.exit(1)
+
+    collection_config = {}
+    if args.get('collection_config'):
+        try:
+            collection_config = validate_collection_config_json(args['collection_config'])
+        except ValidateError:
+            print(
+                f"Validation of provided collection config JSON {args['collection_config']} failed. " +
+                "Please check the formatting of this file.",
```
[the remark this answers is not in the document]
Comment (joverlee521) [reply augur/measurements.py]: Great to not think about spaces. Addressed in https://github.com/nextstrain/augur/pull/879/commits/e4138a31cba8c9885115f2c1905fde397651830f

File: augur/measurements.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (joverlee521) [reply augur/measurements.py]: Addressed in https://github.com/nextstrain/augur/pull/879/commits/e4138a31cba8c9885115f2c1905fde397651830f

File: augur/measurements.py [hunk 13]
```diff
@@ -0,0 +1,364 @@
...
+             "Provided column will be renamed to `value` so please make sure no other columns are named `value`. " +
+             "(default: %(default)s)")
+    required.add_argument("--output-json", required=True, metavar="JSON", type=str,
+        help="Output JSON file. " +
+             "The file name must follow the Auspice sidecar file naming convention to be recognized as a sidecar file. " +
+             "See Nextstrain data format docs for more details.")
+
+    config = export.add_argument_group(
+        title="COLLECTION CONFIGURATION",
+        description="These options control the configuration of the collection for auspice." +
```
[the remark this answers is not in the document]
Comment (joverlee521) [reply augur/measurements.py]: Addressed in https://github.com/nextstrain/augur/pull/879/commits/860d48c5e6c499cde09ab9da6005084540ad3d88

File: augur/measurements.py [hunk 14]
```diff
@@ -0,0 +1,364 @@
...
+    )
+    config.add_argument("--collection-config", metavar="JSON",
+        help="Collection configuration file for advanced configurations. ")
+    config.add_argument("--grouping-column", nargs="+",
+        help="Name of the column(s) that should be used as grouping(s) for measurements. " +
+             "Note that if groupings are provided via command line args, the default group-by " +
+             "field in the config JSON will be dropped.")
+    config.add_argument("--key",
+        help="A short key name of the collection for internal use within auspice." +
+             "If not provided via config or command line option, the collection TSV filename will be used. ")
```
[the remark this answers is not in the document]
Comment (joverlee521) [reply augur/measurements.py]: Addressed in https://github.com/nextstrain/augur/pull/879/commits/860d48c5e6c499cde09ab9da6005084540ad3d88

File: augur/measurements.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (joverlee521) [reply augur/measurements.py]: Changed help message in https://github.com/nextstrain/augur/pull/879/commits/860d48c5e6c499cde09ab9da6005084540ad3d88

File: augur/measurements.py [hunk 15]
```diff
@@ -0,0 +1,364 @@
...
+
+
+    concat = subparsers.add_parser("concat", help="Concatenate multiple measurements JSONs into a single JSON file")
+    concat.add_argument("--jsons", required=True, type=str, nargs="+", metavar="JSONs",
+        help="Measurement JSON files to concatenate.")
+    concat.add_argument("--default-collection", type=str,
+        help="The key of the default collection to display. " +
+             "If not provided, the first collection of the first JSON file will be displayed")
+    concat.add_argument("--minify-json", action="store_true",
+        help="Concat JSONs without indentation or line returns.")
```
[the remark this answers is not in the document]
Comment (joverlee521) [reply augur/measurements.py]: Addressed in https://github.com/nextstrain/augur/pull/879/commits/860d48c5e6c499cde09ab9da6005084540ad3d88

File: augur/measurements.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (joverlee521) [reply augur/measurements.py]: Added global `DEFAULT_ARGS` in https://github.com/nextstrain/augur/pull/879/commits/bd848d92cf31ae27a2cacfff9d8eb96ed3aa61d5

File: augur/measurements.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (joverlee521) [reply augur/measurements.py]: Moved in https://github.com/nextstrain/augur/pull/879/commits/ea31b119cb110b06bf82bae4a2a64fe558bde863

File: augur/validate.py [hunk 16]
```diff
@@ -119,13 +119,175 @@ def export_v1(meta_json, tree_json, **kwargs):
...
+                f"ERROR: Collections at indexes {collection_indexes} share the same collection key '{collection_key}'.",
+                file=sys.stderr
+            )
+
+    # Check the default collection value matches a collection's key value
+    default_collection = measurements.get('default_collection')
+    if default_collection and default_collection not in collection_keys.keys():
+        valid_measurements_config = False
+        print(
+            f"ERROR: The default collection key «{default_collection}» does not match any of the collections' keys.",
```
[the remark this answers is not in the document]
Comment (joverlee521) [reply augur/validate.py]: Addressed in https://github.com/nextstrain/augur/pull/879/commits/e4138a31cba8c9885115f2c1905fde397651830f

File: tests/functional/measurements_concat.t [hunk 17]
```diff
@@ -0,0 +1,42 @@
...
+  {}
+
+Measurements concat for measurements JSONs that have collections that share the same key.
+This is expected to fail.
+
+  $ ${AUGUR} measurements concat \
+  >   --jsons measurements_concat/single_collection_measurements_1.json measurements_concat/single_collection_measurements_1.json \
+  >   --default-collection collection_1 \
+  >   --output-json "$TMP/multiple_collections_measurements.json" &>/dev/null
+  [1]
```
[the remark this answers is not in the document]
Comment (joverlee521) [reply tests/functional/measurements_concat.t]: > I would only write stdout to /dev/null and let stderr get consumed by Cram

Ah, I did not think to do this 🤦‍♀️ . I just redirected everything because I didn't want to include the "Validating..." messages. Addressed in https://github.com/nextstrain/augur/pull/879/commits/bd805bad1a9dd1e9ff29cab826dba2fff306d0dc

File: augur/measurements.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (joverlee521) [reply augur/measurements.py]: Addressed in https://github.com/nextstrain/augur/pull/879/commits/751f13a859a97468638e4fb9dde9a84112823a6c

File: augur/measurements.py [hunk 18]
```diff
@@ -0,0 +1,364 @@
...
+            checks_passed = False
+        # Confirm the provided column does not overwrite an existing column
+        if (required_column in collection_df.columns and
+            provided_column != required_column):
+            print(
+                f"ERROR: Cannot use provided '{provided_column}' column as the {required_column} column " +
+                f"because a '{required_column}' column already exists in collection TSV.",
+                file=sys.stderr,
+            )
+            checks_passed = False
```
[the remark this answers is not in the document]
Comment (joverlee521) [reply augur/measurements.py]: Updated in https://github.com/nextstrain/augur/pull/879/commits/751f13a859a97468638e4fb9dde9a84112823a6c

File: augur/measurements.py [hunk 19]
```diff
@@ -0,0 +1,364 @@
...
+        The collection used to validate groupings
+    grouping_columns: list[str]
+        List of grouping column names
+
+    Returns
+    -------
+    list[dict] or None
+        The groupings for the collection config or None if all grouping columns are invalid
+    """
+    groupings = [{'key': col} for col in grouping_columns if column_exists(collection, col, "grouping")]
```
[the remark this answers is not in the document]
Comment (joverlee521) [reply augur/measurements.py]: Cleared up the code in https://github.com/nextstrain/augur/pull/879/commits/751f13a859a97468638e4fb9dde9a84112823a6c and added error for missing groupings in https://github.com/nextstrain/augur/pull/879/commits/a2f16ddc8af04e73d8cf768587402dda98d7dac7

File: augur/measurements.py [hunk 20]
```diff
@@ -0,0 +1,364 @@
...
+    collection_df = load_collection(args['collection'], args['strain_column'], args['value_column'])
+
+    if collection_df is None:
+        print("ERROR: Loading of collection TSV was unsuccessful. See detailed errors above.", file=sys.stderr)
+        sys.exit(1)
+
+    collection_config = {}
+    if args.get('collection_config'):
+        try:
+            collection_config = validate_collection_config_json(args['collection_config'])
```
[the remark this answers is not in the document]
Comment (joverlee521) [reply augur/measurements.py]: I renamed the alias in https://github.com/nextstrain/augur/pull/879/commits/c6b5091762ed40310b6a8730c25d37a481b27cc2.

File: augur/measurements.py [hunk 21]
```diff
@@ -0,0 +1,364 @@
...
+        help="Show or hide the threshold by default. This will be ignored if no threshold is provided.")
+
+    optional_settings = export.add_argument_group(
+        title="OPTIONAL SETTINGS"
+    )
+    optional_settings.add_argument("--minify-json", action="store_true",
+        help="Export JSON without indentation or line returns.")
+
+
+    concat = subparsers.add_parser("concat", help="Concatenate multiple measurements JSONs into a single JSON file")
```
[the remark this answers is not in the document]
Comment (joverlee521) [reply augur/measurements.py:278]: I took this style of arguments from the `export` modules. Makes the code a little verbose but I agree it looks cleaner in the help messages. Added for concat in https://github.com/nextstrain/augur/pull/879/commits/b903faee8cf79bd5d3af436e842128d211815fd2

File: augur/measurements.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (joverlee521) [reply augur/measurements.py]: Addressed in https://github.com/nextstrain/augur/pull/879/commits/c8a2e0ae82625864eb527c5c6db5dfea4b3d4666

File: augur/measurements.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (joverlee521) [reply augur/measurements.py]: Added the `--include-columns` option in https://github.com/nextstrain/augur/pull/879/commits/cf9dfbd1aef5f76816550f02512a496aaae47ae6

File: augur/measurements.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (joverlee521) [reply augur/measurements.py]: Made the overhaul commit https://github.com/nextstrain/augur/pull/879/commits/d22aeeec1e3fdebf5a758e6331cdb3bd00958d6a. Let me know if this looks better than splitting into functions!

[the remark this answers is not in the document]
Comment (huddlej) [reply augur/measurements.py]: Thanks, @joverlee521! d22aeee looks great to me, although I should clarify my original comment.

I do think functions are great for abstracting and encapsulating logic, even when that logic is only used once or never exposed through a public API. For example, [moving augur filter's logic into separate functions with a consistent interface](https://github.com/nextstrain/augur/commit/0710b500ee6f50f090636e9858fd3dc86a327952) made their internal logic of each filter easier to check and test, the calling code clearer to read, and new filters easier to add.

My main concern in the comment above was that the functions still behaved as those they were part of the "main" CLI logic by interacting directly with the operating system through print statements and sys exit calls. Those types of interactions are difficult to handle properly in the calling code. For example, if we decided to handle all exceptions of a specific kind at the top-level of Augur (like @victorlin [recently did for FileNotFoundError exceptions](https://github.com/nextstrain/augur/pull/914/files)), we would like to know our functions will raise these exceptions to be handled downstream like that instead of trying to handle the errors internally.

If you preferred to keep the logic from this PR split out into functions, we would just need those functions to handle errors differently than the main CLI function does by raising exceptions.

[the remark this answers is not in the document]
Comment (joverlee521) [reply augur/measurements.py]: I'm actually pretty happy with the overhaul. It's more straightforward!

**Comment by joverlee521** [comment]
Rebased on master and fixed up some small fix commits. I'll merge after all of the CI tests have passed.

**Review by huddlej — APPROVED** [review]
